File: 1.two_sum/solution.cpp
#include <iostream>
#include <vector>
#include <map>
#include <memory>
using namespace std;


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        typedef unsigned int Index;
        typedef int Diff;
        vector<int> dft;
        
        if (nums.size() < 2) {
            return dft;
        }
        map<Diff, Index> _temp;
        for (size_t pos = 0; pos < nums.size(); pos++) {
            if (auto itr = _temp.find(target - nums[pos]); 
                itr != _temp.end()) {
                return vector<int>{static_cast<int>(itr->second), 
                static_cast<int>(pos)};
            } else {
                _temp[nums[pos]] = pos;
            }
        }
        return std::move(dft);
    }
};


int main() {
    Solution s;
    vector<int> input{2, 7, 11, 15};
    for (auto & i : s.twoSum(input, 9)) {
        cout<<i<<std::endl;
    }
}

File: 1.two_sum/solution.go
package main

import "sort"

func twoSum_hash(nums []int, target int) []int {
	m := map[int]int{}
	for i, cur := range nums {
		if pre, ok := m[target-cur]; ok {
			return []int{pre, i}
		} else {
			m[cur] = i
		}
	}
	return []int{}
}

func twoSum(nums []int, target int) []int {
	index := map[int]int{}
	for i, v := range nums {
		index[v] = i
	}
	sort.Ints(nums)
	for l, r := 0, len(nums)-1; l <= r; {
		sum := nums[l] + nums[r]
		if sum == target {
			return []int{index[nums[l]], index[nums[r]]}
		} else if sum > target {
			r--
		} else {
			l++
		}
	}
	return []int{}
}
